### PR TITLE
move lotus mount, dag store etc from markets to lotus

### PR DESCRIPTION
**Pull request by dirkmc**

### CODE REVIEW

File: markets/dagstore/mount_api.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+			if err == nil {
+				return reader, nil
+			}
+		}
+	}
+
+	lastErr := xerrors.New("no sectors found to unseal from")
+	// if there is no unsealed sector containing the piece, just read the piece from the first sector we are able to unseal.
+	for _, deal := range pieceInfo.Deals {
+		reader, err := m.rm.UnsealSector(ctx, deal.SectorID, deal.Offset.Unpadded(), deal.Length.Unpadded())
```
Comment (raulk) [inline markets/dagstore/mount_api.go:61]: Can we add a note that this will block for a long time with the current PoRep?

File: markets/dagstore/mount_api_test.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			require.NoError(t, err)
+			bz, err := io.ReadAll(r)
+			require.NoError(t, err)
+
+			require.Equal(t, tc.fetchedData, string(bz))
+		})
+	}
+}
+
+func TestLotusMountApiGetUnpaddedCARSize(t *testing.T) {
```
Comment (raulk) [inline markets/dagstore/mount_api_test.go:96]: reminder to rename tests if the above suggestion of rename this component to `LotusAccessor` is accepted!

File: node/modules/storageminer.go [hunk 3]
```diff
@@ -609,9 +609,10 @@ func StorageProvider(minerAddress dtypes.MinerAddress,
 	}
 
 	opt := storageimpl.CustomDealDecisionLogic(storageimpl.DealDeciderFunc(df))
+	shardMigrator := storageimpl.NewShardMigrator(address.Address(minerAddress), ds, dagStore, pieceStore, spn)
```
Comment (raulk) [inline node/modules/storageminer.go:612]: Can we move the shard migrator here too? It too is specific to Lotus, so I think it makes sense to move into the Lotus repo.

(Suggestion: submit another PR on top of this one to avoid bloating this one)

File: markets/dagstore/dagstorewrapper.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
+package dagstore
```
Comment (raulk) [inline markets/dagstore/dagstorewrapper.go:1]: Rename this file to `wrapper.go`, since `dagstore` is already part of the folder and package?

File: node/modules/storageminer.go [hunk 5]
```diff
@@ -675,7 +676,7 @@ func RetrievalProvider(
 	dt dtypes.ProviderDataTransfer,
 	pricingFnc dtypes.RetrievalPricingFunc,
 	userFilter dtypes.RetrievalDealFilter,
-	dagStore mktdagstore.DagStoreWrapper,
+	dagStore shared.DagStoreWrapper,
```
Comment (raulk) [inline node/modules/storageminer.go:679]: Let's avoid using _their_ interface here. Instead, just wire in _our_ concrete type `dagstore.Wrapper`, which will be present in the DI graph, and the `NewProvider` constructor will accept because it implements the interface _they_ want.

nit: I'm not a fan of shared/util/helpers/etc. package names in Go. It's actually [quite frowned upon](https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common). ¯\_(ツ)_/¯

File: markets/dagstore/dagstorewrapper.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		Datastore:     cfg.Datastore,
+		MountRegistry: registry,
+		FailureCh:     failureCh,
+	}
+	dagStore, err := dagstore.NewDAGStore(dcfg)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to create DAG store: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (dirkmc) [reply markets/dagstore/dagstorewrapper.go]: Added Start and Stop hooks that call a new `Start(ctx)` method and the existing `Close()` method respectively

File: markets/dagstore/dagstorewrapper.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	defer ds.wg.Done()
+	ticker := time.NewTicker(gcInterval)
+	defer ticker.Stop()
+
+	select {
+	case <-ticker.C:
+		_, _ = ds.dagStore.GC(ds.ctx)
+	case f := <-failureCh:
+		log.Errorw("shard failed", "shard-key", f.Key.String(), "error", f.Error)
+		if err := ds.dagStore.RecoverShard(ds.ctx, f.Key, nil, dagstore.RecoverOpts{}); err != nil {
```
[the remark this answers is not in the document]
Comment (dirkmc) [reply markets/dagstore/dagstorewrapper.go]: I added a `chan ShardResult` but the resulting code is very complex. I think we may need to revisit the pattern of passing a channel to the DAG store methods, as the calling code needs to keep track of how many times it's called the method and drain the channel afterwards, while ensuring there are no races.

File: markets/dagstore/dagstorewrapper.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	err := ds.dagStore.AcquireShard(ctx, key, resch, dagstore.AcquireOpts{})
+
+	if err != nil {
+		if xerrors.Unwrap(err) != dagstore.ErrShardUnknown {
+			return nil, xerrors.Errorf("failed to schedule acquire shard for piece CID %s: %w", pieceCid, err)
+		}
+
+		// if the DAGStore does not know about the Shard -> register it and then try to acquire it again.
+		log.Warnw("failed to load shard as shard is not registered, will re-register", "pieceCID", pieceCid)
+		if err := shared.RegisterShardSync(ctx, ds, pieceCid, "", false); err != nil {
```
[the remark this answers is not in the document]
Comment (dirkmc) [reply markets/dagstore/dagstorewrapper.go]: @aarshkshah1992 can you please comment here, I'm not sure how this works 🙏

[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply markets/dagstore/dagstorewrapper.go]: @dirkmc 

It's the path of a transient file that we can ask the DAG Store to use to perform the Indexing rather than fetching it via the Mount if we already have a transient file. However, we don't have it here and therefor we pass an empty file path.

File: markets/dagstore/dagstorewrapper.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/dagstore/mount"
+	"github.com/filecoin-project/dagstore/shard"
+	"github.com/filecoin-project/go-fil-markets/carstore"
+	"github.com/filecoin-project/go-fil-markets/shared"
+)
+
+var log = logging.Logger("dagstore-wrapper")
+var gcInterval = 5 * time.Minute
```
[the remark this answers is not in the document]
Comment (dirkmc) [reply markets/dagstore/dagstorewrapper.go]: Sounds good 👍 
Suggest we do this as a separate PR

**Comment by aarshkshah1992** [comment]
@raulk Merging it as per today's discussion so we can start testing. Please continue leaving comments here and we'll make sure we address them.
